### ISSUE TITLE
Using Conda to install Galaxy framework dependencies 

### ISSUE THREAD

**Issue by sveinugu** [galaxyproject/galaxy#3876]
I talked to @bgruening some weeks ago whether it was possible to install Galaxy framework dependencies using Conda, and he informed me that this is still not possible, but that @jmchilton might be working on it. There might be an issue for this already, but I did not manage to find it.

How I understand the current situation:
- If Conda is available at startup, Galaxy uses pip in Conda's Python instead of Galaxy pip. The former  does not include pre-built binaries, so these needs to be compiled (e.g. for packages like Numpy). (https://docs.galaxyproject.org/en/master/admin/framework_dependencies.html#conda)
- However, if one were to support using Conda directly to install packages, they will (in my understanding) come pre-built (e.g. Numpy prebuilt to support the Intel MKL library). 

How @bgruening envisioned it would be to add another requirements file, alongside the pip requirements, for the Conda packages to install.

I am quite sure that this feature would be very helpful for standard Galaxy installations.

Where I am coming from: Supporting Conda in the framework installation would be very helpful for the Genomic HyperBrowser deployment, which is dependent on (among others) R. Automatically installing R in the correct version would make it much easier to deploy. Also, we would like to support using the Intel MKL library out-of-the-box.

**Comment by jmchilton** [comment]
I'm not actively working on this. @natefoo would probably be a better person to ping on this.

I'd say we are pretty close to being able to support this but there are a few things I think need to happen. The thing I'd like most to see resolved since it would help installing Galaxy from pypi as well is to figure out what to do about pysam. We use an older forked version of that project so we either need to publish our fork (probably with a different name?) to PyPI and Conda (probably a new channel just for us) or we need to get our changes merged back in.  I believe we are using this branch here - https://github.com/dannon/pysam/tree/flexindex_84.

After that I think it would be best to move to a meta-requirements file that builds all of our particular requirements file. We want to support pinned PyPI, pinned and hashed PyPI, unpinned PyPI, and Conda requirements files - all of which are related but different and incompatible. So we should have a master YAML file and generate the particular requirements from that. We've talked about this issue and I think people discussing it were on board for a metafile this way - but I haven't done any research to determine if such a format already exists and can be leveraged. (It wouldn't be complicated to write at all - but if there were some tooling it might be worth leveraging it.)

Without figuring out pysam we are going to be one custom dependency away from Conda support and without the meta-file it would be too onerous to ask people changing dependencies to change a bunch of different files I think. With those two changes however - Conda support should be pretty straightforward and probably would already work.

**Comment by natefoo** [comment]
+1 to what @jmchilton said.

**Comment by sveinugu** [comment]
@natefoo Any news on this? Will it be actively worked on? Or will this require a bunch of more +1 s? :-)

**Comment by mvdbeek** [comment]
This has worked for a while now. The latest and greatest instructions for 19.01 are at https://docs.galaxyproject.org/en/latest/admin/framework_dependencies.html#conda.

This uses conda and a conda-compatible virtualenv to install the same dependencies you'd get on a normal setup, which gives you the versions that we run our tests against. I have been using this in production for a while and it works well.

**Comment by gregvonkuster** [comment]
Sorry to post to a closed issue, but not sure if this warrants a new issue or not.  I recently switched over from the Python virtualenv to the conda env and found that I had to comment out this line in the common_startup.sh file https://github.com/galaxyproject/galaxy/blob/dev/scripts/common_startup.sh#L129.  Leaving it as is would attempt to use virtualenv which is not found.  Has anyone else seen this or is it a weird issue at my end?

**Comment by mvdbeek** [comment]
You always, really always, want to run galaxy in  a virtualenv, even if you're using Conda to provide the python interpreter. If you don't do this you will very likely run into subtle but hard to debug problems.

**Comment by gregvonkuster** [comment]
So I obviously missed installing virtualenv - do I simply "conda install virtualenv" or something similar before starting Galaxy?  Thanks for the help!

**Comment by mvdbeek** [comment]
This is explained in https://docs.galaxyproject.org/en/latest/admin/framework_dependencies.html#conda for 19.01 or newer. So just run.sh with the conda binary on PATH should be enough to set up a virtualenv in .venv.

**Comment by gregvonkuster** [comment]
Hmm...those were the instructions I used, but "virtualenv" is not found on my path.  I'll work this out offline though.  Thanks @mvdbeek for help in confirming that I have some kind of issue in my environment.

**Comment by mvdbeek** [comment]
I can double check this, we did some small modifications to the code after writing the instructions.

**Comment by mvdbeek** [comment]
Yeah, this is indeed broken (at least if using zsh):
```
 /tmp  (ssh)  git clone --depth 50 -b release_19.01 https://github.com/galaxyproject/galaxy  && cd galaxy && sh run.sh       SIGHUP(1) ↵  7009  14:44:11
Cloning into 'galaxy'...
remote: Enumerating objects: 49172, done.
remote: Counting objects: 100% (49172/49172), done.
remote: Compressing objects: 100% (25220/25220), done.
remote: Total 49172 (delta 34045), reused 34963 (delta 23142), pack-reused 0
Receiving objects: 100% (49172/49172), 83.80 MiB | 165.00 KiB/s, done.
Resolving deltas: 100% (34045/34045), done.
Initializing config/migrated_tools_conf.xml from migrated_tools_conf.xml.sample
Initializing config/shed_tool_conf.xml from shed_tool_conf.xml.sample
Initializing config/shed_tool_data_table_conf.xml from shed_tool_data_table_conf.xml.sample
Initializing config/shed_data_manager_conf.xml from shed_data_manager_conf.xml.sample
Initializing tool-data/shared/ucsc/builds.txt from builds.txt.sample
Initializing tool-data/shared/ucsc/manual_builds.txt from manual_builds.txt.sample
Initializing static/welcome.html from welcome.html.sample
Found conda at: /home/mvandenb/miniconda3/bin/conda
Found Conda, will set up a virtualenv using conda.
To use a virtualenv instead, create one with a non-Conda Python 2.7 at .venv
Creating Conda environment for Galaxy: _galaxy_
To avoid this, use the --no-create-venv flag or set $GALAXY_CONDA_ENV to an
existing environment before starting Galaxy.
Collecting package metadata: done
Solving environment: done

## Package Plan ##

  environment location: /home/mvandenb/miniconda3/envs/_galaxy_

  added / updated specs:
    - pip[version='>=9']
    - python=2.7
    - virtualenv[version='>=16']


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2018.11.29         |        py27_1000         145 KB  conda-forge
    libffi-3.2.1               |    hf484d3e_1005          46 KB  conda-forge
    libgcc-ng-7.3.0            |       hdf63c60_0         6.1 MB  conda-forge
    libstdcxx-ng-7.3.0         |       hdf63c60_0         2.6 MB  conda-forge
    ncurses-6.1                |    hf484d3e_1002         1.3 MB  conda-forge
    openssl-1.0.2p             |    h14c3975_1002         3.1 MB  conda-forge
    pip-19.0.2                 |           py27_0         1.8 MB  conda-forge
    python-2.7.15              |    h938d71a_1006        11.9 MB  conda-forge
    readline-7.0               |    hf8c457e_1001         391 KB  conda-forge
    setuptools-40.8.0          |           py27_0         630 KB  conda-forge
    sqlite-3.26.0              |    h67949de_1000         1.9 MB  conda-forge
    tk-8.6.9                   |    h84994c4_1000         3.1 MB  conda-forge
    virtualenv-16.0.0          |        py27_1000         1.8 MB  conda-forge
    wheel-0.33.0               |           py27_0          34 KB  conda-forge
    zlib-1.2.11                |    h14c3975_1004         101 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        35.0 MB

The following NEW packages will be INSTALLED:

  ca-certificates    conda-forge/linux-64::ca-certificates-2018.11.29-ha4d7672_0
  certifi            conda-forge/linux-64::certifi-2018.11.29-py27_1000
  libffi             conda-forge/linux-64::libffi-3.2.1-hf484d3e_1005
  libgcc-ng          conda-forge/linux-64::libgcc-ng-7.3.0-hdf63c60_0
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-7.3.0-hdf63c60_0
  ncurses            conda-forge/linux-64::ncurses-6.1-hf484d3e_1002
  openssl            conda-forge/linux-64::openssl-1.0.2p-h14c3975_1002
  pip                conda-forge/linux-64::pip-19.0.2-py27_0
  python             conda-forge/linux-64::python-2.7.15-h938d71a_1006
  readline           conda-forge/linux-64::readline-7.0-hf8c457e_1001
  setuptools         conda-forge/linux-64::setuptools-40.8.0-py27_0
  sqlite             conda-forge/linux-64::sqlite-3.26.0-h67949de_1000
  tk                 conda-forge/linux-64::tk-8.6.9-h84994c4_1000
  virtualenv         conda-forge/linux-64::virtualenv-16.0.0-py27_1000
  wheel              conda-forge/linux-64::wheel-0.33.0-py27_0
  zlib               conda-forge/linux-64::zlib-1.2.11-h14c3975_1004



Downloading and Extracting Packages
libstdcxx-ng-7.3.0   | 2.6 MB    | ################################################################################################################### | 100%
setuptools-40.8.0    | 630 KB    | ################################################################################################################### | 100%
readline-7.0         | 391 KB    | ################################################################################################################### | 100%
sqlite-3.26.0        | 1.9 MB    | ################################################################################################################### | 100%
ncurses-6.1          | 1.3 MB    | ################################################################################################################### | 100%
openssl-1.0.2p       | 3.1 MB    | ################################################################################################################### | 100%
wheel-0.33.0         | 34 KB     | ################################################################################################################### | 100%
zlib-1.2.11          | 101 KB    | ################################################################################################################### | 100%
libgcc-ng-7.3.0      | 6.1 MB    | ################################################################################################################### | 100%
certifi-2018.11.29   | 145 KB    | ################################################################################################################### | 100%
pip-19.0.2           | 1.8 MB    | ################################################################################################################### | 100%
tk-8.6.9             | 3.1 MB    | ################################################################################################################### | 100%
virtualenv-16.0.0    | 1.8 MB    | ################################################################################################################### | 100%
python-2.7.15        | 11.9 MB   | ################################################################################################################### | 100%
libffi-3.2.1         | 46 KB     | ################################################################################################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use:
# > source activate _galaxy_
#
# To deactivate an active environment, use:
# > source deactivate
#

Activating Conda environment: _galaxy_
WARNING: Your shell is not supported with Conda, attempting to use Conda env
         '_galaxy_' with manual environment setup. To avoid this
         message, use a supported shell or activate the environment before
         starting Galaxy.
New python executable in /tmp/galaxy/.venv/bin/python
Installing setuptools, pip, wheel...
  Complete output from command /tmp/galaxy/.venv/bin/python - setuptools pip wheel:
  Traceback (most recent call last):
  File "<stdin>", line 11, in <module>
ImportError: cannot import name main
----------------------------------------
...Installing setuptools, pip, wheel...done.
Traceback (most recent call last):
  File "/home/mvandenb/miniconda3/envs/_galaxy_/bin/virtualenv", line 11, in <module>
    sys.exit(main())
  File "/home/mvandenb/.local/lib/python2.7/site-packages/virtualenv.py", line 712, in main
    symlink=options.symlink)
  File "/home/mvandenb/.local/lib/python2.7/site-packages/virtualenv.py", line 947, in create_environment
    download=download,
  File "/home/mvandenb/.local/lib/python2.7/site-packages/virtualenv.py", line 904, in install_wheel
    call_subprocess(cmd, show_stdout=False, extra_env=env, stdin=SCRIPT)
  File "/home/mvandenb/.local/lib/python2.7/site-packages/virtualenv.py", line 796, in call_subprocess
    % (cmd_desc, proc.returncode))
OSError: Command /tmp/galaxy/.venv/bin/python - setuptools pip wheel failed with error code 1
```

**Comment by mvdbeek** [comment]
I think https://github.com/galaxyproject/galaxy/pull/7360 should be more portable and does the right thing for me. @gregvonkuster can you give this a try ?

**Comment by gregvonkuster** [comment]
Ah, that solved my issue.  I was using ```conda activate _galaxy_```, which threw the exception.  So I'll switch to ```. ~conda/bin/activate ~/conda/envs/_galaxy_``` which works!  Thanks a bunch @mvdbeek !  ;)

**Comment by mvdbeek** [comment]
Glad that works, but what I meant was that you'd just put the `conda` binary on PATH (if it isn't already). You don't need to activate `_galaxy_` manually, though that doesn't hurt either.

**Comment by gregvonkuster** [comment]
Yeah, conda actually is on my path, I just used the full paths in my comments for clarity to other readers.  My understanding is that I should actually be manually activating the _galaxy_ conda env before starting Galaxy though.  Is that not correct?

**Comment by mvdbeek** [comment]
> My understanding is that I should actually be manually activating the _galaxy_ conda env before starting Galaxy though

Just during the first time running Galaxy, afterwards this isn't necessary.
